src/main/java/com/microsoft/sqlserver/jdbc/SQLServerPreparedStatement.java

```diff
@@ -967,38 +995,57 @@ private boolean doPrepExec(TDSWriter tdsWriter,
             Parameter[] params,
             boolean hasNewTypeDefinitions,
             boolean hasExistingTypeDefinitions) throws SQLServerException {
-        
+
         boolean needsPrepare = (hasNewTypeDefinitions && hasExistingTypeDefinitions) || !hasPreparedStatementHandle();
 
         // Cursors don't use statement pooling.
         if (isCursorable(executeMethod)) {
-            
-            if (needsPrepare) 
+            if (needsPrepare)
                 buildServerCursorPrepExecParams(tdsWriter);
             else
                 buildServerCursorExecParams(tdsWriter);
         }
         else {
+            int paramCount = SQLServerConnection.countParams(userSQL);
+            if (needsPrepare && null != procedureName && paramCount != 0 && !isTVPType(params)) { 
```

Comment:
A few questions here:
1. How do we know that this is an RPC call with separate list of parameters vs. something like "EXEC spTobias @dobidoo = 2" as a string plus separately passing other parameters.
2. We should support this also for 1 == paramCount, no?

Reply:
Good question:
for queries like `EXEC spTobias @dobidoo = 2` the param count is still counted as 0. The countParams(userSQL) method will return 0. This method which was already used in the driver, counts params by the number of ? marks.  (since we are not using queries such as `EXEC spTobias (?)` and then later bind the parameters. For such queries for now, we use the current behavior of driver which is 'sp_execute'. 

I think that would require the countParams to parse the queries for parameters passed in userSQL too. 
I will try to add this with the next push.

Reply:
To clarify, how do you handle this "EXEC spTobias 2, ?", where paramCount is 1?

Reply:
You are correct @TobiasSQL .. Good catch.. in that case we need to check the query for such scenarios. One solution is to look for '? ' marks and replace them with parameters that is passed. Instead of RPC call we can directly use a query call.
For now, I added a check for those queries and not use rpc calls, instead use the old behavior of calling sp_executesql .
What do you think about this?

Reply:
How do you detect this (i.e. that we are either strictly calling SP without params or only ?s) ? I couldn't find it.

Reply:
@TobiasSQL Thank you so much for following up. please check this method on callRPCDirectly in [here](https://github.com/Microsoft/mssql-jdbc/pull/547/files#diff-a51517a0e4bf7e029ed04f9476a37f89R1042) and its implementation in [here](https://github.com/Microsoft/mssql-jdbc/pull/547/files#diff-0a4455f5ffd8e31e9e1badaa7822375eR284).
Basically by counting the number of parameters (?) and commas. 
If there is no ? (meaning parameters in the query) or there sql is something like `exec proc ?, 1, 2` we directly use sp_executesql.
Do you think it makes sense?